### ISSUE TITLE
Perl : 5.34 test failures with Inf NaN numbers on Windows

### ISSUE THREAD

**Issue by jameysiddiqui** [Perl/perl5#19010]
CCTYPE = MSVC120

Failed test 359 - sprintf %c stringy NaN undef at t/op/infnan.t line 31
Failed test 360 - stringy NaN sprintf %c fails at t/op/infnan.t line 32
got ''
expected /(?^:Cannot printf)/
Failed test 363 - chr stringy NaN undef at t/op/infnan.t line 325
Failed test 364 - stringy NaN chr() fails at t/op/infnan.t line 326
got ''

Please let me know what needs to be done
expected /(?^:Cannot chr)/
Failed test 368 - stringy NaN pack c fails at t/op/infnan.t line 334
got 'Cannot pack 0 with 'c' at t/op/infnan.t line 332.\n'
expected /(?^:Cannot pack NaN)/
Failed test 371 - pack C stringy NaN undef at t/op/infnan.t line 332
Failed test 372 - stringy NaN pack C fails at t/op/infnan.t line 334
got ''
expected /(?^:Cannot pack NaN)/
Failed test 375 - pack s stringy NaN undef at t/op/infnan.t line 332
Failed test 376 - stringy NaN pack s fails at t/op/infnan.t line 334
got ''

**Comment by jkeenan** [comment]
> CCTYPE = MSVC120
> 
> Failed test 359 - sprintf %c stringy NaN undef at t/op/infnan.t line 31
> Failed test 360 - stringy NaN sprintf %c fails at t/op/infnan.t line 32
> got ''
> expected /(?^:Cannot printf)/
> Failed test 363 - chr stringy NaN undef at t/op/infnan.t line 325
> Failed test 364 - stringy NaN chr() fails at t/op/infnan.t line 326
> got ''
> 
> Please let me know what needs to be done
> expected /(?^:Cannot chr)/
> Failed test 368 - stringy NaN pack c fails at t/op/infnan.t line 334
> got 'Cannot pack 0 with 'c' at t/op/infnan.t line 332.\n'
> expected /(?^:Cannot pack NaN)/
> Failed test 371 - pack C stringy NaN undef at t/op/infnan.t line 332
> Failed test 372 - stringy NaN pack C fails at t/op/infnan.t line 334
> got ''
> expected /(?^:Cannot pack NaN)/
> Failed test 375 - pack s stringy NaN undef at t/op/infnan.t line 332
> Failed test 376 - stringy NaN pack s fails at t/op/infnan.t line 334
> got ''

Do you know whether these failures also appeared in perl-5.32.0 on this platform?

**Comment by jameysiddiqui** [comment]
yes 5.32 also failing only 5.28 is passing, I wonder how the perl is marked stable without running tests on windows

**Comment by sisyphus** [comment]
Most likely it's the compiler that @jameysiddiqui is using that's creating the difference.
I believe that "MSVC120" is also known as Visual C++ 2013 or Visual C++ 12.0.

For me, MSVC142 worked fine for building 5.32.0.
@jameysiddiqui,  is switching to that tool chain an acceptable option for you ?
I haven't actually tested that compiler with perl-5.34.0, but I will do so tonight.

I don't have MSVC120 at hand. (All of my other MS tool chains are pre-MSVC110.)

@jameysiddiqui, does the same failure occur with 5.30.0 ? That will help us narrow it down a bit more.
Also, we probably should take a look at your<code>perl -V</code> output.
If you haven't actually installed that perl, just <code>cd</code> to the top level perl source directory of your build and run:
```
.\perl -I.\lib -V
```
and then post that output.

On the basis of the report, it looks to me that things are fine with this perl's handling of NaNs, except that:

1) Our expectation is that <code>sprintf("%c", "NaN")</code>will numify the string "NaN" to a NaN, but this expectation is not met;

2) Our expectation is that <code>chr("NaN")</code> will numify the string "NaN" to a NaN , but this expectation is not met;

3) Our expectation is that, when $f is either 'c' or 'C' or 's', <code>pack($f, "NaN")</code> will numify the string "NaN" to a NaN , but this expectation is not met.

I think that in all 3 cases, the string "NaN" remains a string, though I'm not sure how that fits in with the "Cannot pack 0 with 'c' ...." error message in relation to test 368.

Is that the full extent of the failing tests ?
I would have thought other pack() templates (eg 'S') would have thrown up additional test failures for this particular build.

**Comment by jameysiddiqui** [comment]
This is full test of failures with perl 5.34 on windows with CCTYPE = MSVC120

*********************************************************************

nmake test failed on windows

Test Summary Report
-------------------
**re/uniprops03.t (Wstat: 65280 Tests: 38532 Failed: 12)**
Failed tests: 6859-6862, 6889-6892, 6925-6928
Non-zero exit status: 255
**opbasic/cmp.t (Wstat: 0 Tests: 13280 Failed: 9)**
Failed tests: 49, 637, 1211, 1841, 1855, 1869, 1883, 1897
1911
**op/infnan.t (Wstat: 0 Tests: 1106 Failed: 41)**
Failed tests: 359-360, 363-364, 368, 371-372, 375-376
379-380, 383-384, 387-388, 391-392, 395-396
399-400, 403-404, 407-408, 411-412, 415-416
419-420, 423-424, 427-428, 431-432, 435-436
439-440
**op/time.t (Wstat: 0 Tests: 72 Failed: 2)**
Failed tests: 71-72
../dist/Net-Ping/t/450_service.t (Wstat: 0 Tests: 26 Failed: 0)
TODO passed: 9
../ext/IPC-Open3/t/IPC-Open3.t (Wstat: 0 Tests: 45 Failed: 0)
TODO passed: 25
**../lib/warnings.t (Wstat: 0 Tests: 923 Failed: 1)**
Failed test: 694
Files=2677, Tests=1131662, 1653 wallclock secs (101.27 usr + 9.02 sys = 110.28 CPU)
Result: FAIL
NMAKE : fatal error U1077: '.\perl.exe' : return code '0x41'
Stop.
*****************************************************************************************************
The issue is happening with perl 5.30 as well
This is output of  **"perl -I.\lib -V"**

C:\Jamey\perlbackup\perl-5.30.3.tar\perl-5.30.3>perl -I.\lib -V
Summary of my perl5 (revision 5 version 30 subversion 3) configuration:

  Platform:
    osname=MSWin32
    osvers=6.3.9600
    archname=MSWin32-x64-multi-thread
    uname=''
    config_args='undef'
    hint=recommended
    useposix=true
    d_sigaction=undef
    useithreads=define
    usemultiplicity=define
    use64bitint=define
    use64bitall=undef
    uselongdouble=undef
    usemymalloc=n
    default_inc_excludes_dot=define
    bincompat5005=undef
  Compiler:
    cc='cl'
    ccflags ='-nologo -GF -W3 -O1 -MD -Zi -DNDEBUG -GL -fp:precise -DWIN32 -D_CONSOLE -DNO_STRICT -DWIN64 -DCONSERVATIVE -D_CRT_SECURE_NO_DEPRECATE -D
_CRT_NONSTDC_NO_DEPRECATE  -DPERL_TEXTMODE_SCRIPTS -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS'
    optimize='-O1 -MD -Zi -DNDEBUG -GL -fp:precise'
    cppflags='-DWIN32'
    ccversion='18.00.40629'
    gccversion=''
    gccosandvers=''
    intsize=4
    longsize=4
    ptrsize=8
    doublesize=8
    byteorder=12345678
    doublekind=3
    d_longlong=undef
    longlongsize=8
    d_longdbl=define
    longdblsize=8
    longdblkind=0
    ivtype='__int64'
    ivsize=8
    nvtype='double'
    nvsize=8
    Off_t='__int64'
    lseeksize=8
    alignbytes=8
    prototype=define
  Linker and Libraries:
    ld='link'
    ldflags ='-nologo -nodefaultlib -debug -opt:ref,icf -ltcg  -libpath:"c:\perl\lib\CORE"  -machine:AMD64 -subsystem:console,"5.02"'
    libpth="C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\\lib\amd64"
    libs=oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib uuid.
lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
    perllibs=oldnames.lib kernel32.lib user32.lib gdi32.lib winspool.lib  comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib  netapi32.lib u
uid.lib ws2_32.lib mpr.lib winmm.lib  version.lib odbc32.lib odbccp32.lib comctl32.lib msvcrt.lib
    libc=msvcrt.lib
    so=dll
    useshrplib=true
    libperl=perl530.lib
    gnulibc_version=''
  Dynamic Linking:
    dlsrc=dl_win32.xs
    dlext=dll
    d_dlsymun=undef
    ccdlflags=' '
    cccdlflags=' '
    lddlflags='-dll -nologo -nodefaultlib -debug -opt:ref,icf -ltcg  -libpath:"c:\perl\lib\CORE"  -machine:AMD64 -subsystem:console,"5.02"'


Characteristics of this binary (from libperl):
  Compile-time options:
    HAS_TIMES
    HAVE_INTERP_INTERN
    MULTIPLICITY
    PERLIO_LAYERS
    PERL_COPY_ON_WRITE
    PERL_DONT_CREATE_GVSV
    PERL_IMPLICIT_CONTEXT
    PERL_IMPLICIT_SYS
    PERL_MALLOC_WRAP
    PERL_OP_PARENT
    PERL_PRESERVE_IVUV
    USE_64_BIT_INT
    USE_ITHREADS
    USE_LARGE_FILES
    USE_LOCALE
    USE_LOCALE_COLLATE
    USE_LOCALE_CTYPE
    USE_LOCALE_NUMERIC
    USE_LOCALE_TIME
    USE_PERLIO
    USE_PERL_ATOF
    USE_THREAD_SAFE_LOCALE
  Built under MSWin32
  Compiled at Aug  1 2021 16:03:10
  @INC:
    .\lib
    C:/Jamey/perlbackup/perl-5.30.3.tar/perl-5.30.3/lib
***************************************************************************************************************************

I have not tried with MSVC14, I have to try yet..

**Comment by sisyphus** [comment]
I'm wondering if <code>strtod("NaN", NULL)</code> might be posing a problem on MSVC120.
Firstly, could you verify that <code>perl -V:d_strtod</code> reports <code>d_strtod='define';</code>.
(The <code>perl -V</code> output doesn't tell us the status of <code>d_strtod</code>.)

Assuming that <code>d_strtod</code> is defined, could you build the following simple C program and tell us what it outputs:
```
#include <stdio.h>
#include <stdlib.h>

int main(void) {
 printf("%f\n", strtod("NaN", NULL));
 return 0;
}

```
With MSVC142, it simply outputs "nan". Is it the same with MSVC120 ?

Incidentally, MSVC142 builds both perl-5.32.0 and perl-5.34.0 ok for me - no problems at all with t/op/infnan.t

**Comment by jameysiddiqui** [comment]
Output
MSVC120: 0.000000 
MSVC141: nan
MSVS142: nan
I will check building perl with MSVC141/2

**Comment by sisyphus** [comment]
My old Windows Server 2003 SP1 Platform SDK compiler (_MSC_VER  is 1400) also has <code>strtod("NaN", NULL);</code> return  0.000000.
But unlike your MSVC120, it builds perl-5.30.0 just fine - with all tests passing.

It won't build perl-5.34.0 at all, producing the following error (which I haven't investigated) during the <code>nmake</code> step:
```
Running pm_to_blib for cpan/Pod-Escapes directly
Generating a nmake-style Makefile
Writing Makefile for Pod::Functions
        "C:\_64\comp-sdk\perl-5.34.0\miniperl.exe" "-I..\..\lib" Functions_pm.PL ..\..\pod\perlfunc.pod
No types listed for accept
NMAKE : fatal error U1077: '"C:\_64\comp-sdk\perl-5.34.0\miniperl.exe"' : return code '0x19'
Stop.
        "C:\_64\comp-sdk\perl-5.34.0\miniperl.exe" "-I..\..\lib" Functions_pm.PL ..\..\pod\perlfunc.pod
No types listed for accept
NMAKE : fatal error U1077: '"C:\_64\comp-sdk\perl-5.34.0\miniperl.exe"' : return code '0x19'
Stop.
Unsuccessful make(ext/Pod-Functions): code=512 at ..\make_ext.pl line 584.
NMAKE : fatal error U1077: '..\miniperl.exe' : return code '0x2'
Stop.
```
However, when it comes to building perl-5.32.0, this Windows Server 2003 SP1 Platform SDK compiler produces exactly the same     
 errors as the ones in the failing "Test Summary Report" that you provided.

So, I guess this at least gives me a chance of working out what's going on, and maybe even how to fix it.
(It might take me a while.)

**Comment by sisyphus** [comment]
The following patch to the perl-5.32.0/numeric.c fixes the opbasic/cmp.t, op/infnan.t, op/time.t and lib/warnings.t failures for me:
```
--- numeric.c_orig      2021-08-03 18:28:21 +1000
+++ numeric.c   2021-08-03 19:30:11 +1000
@@ -822,10 +822,11 @@
             s++;

             flags |= IS_NUMBER_NAN | IS_NUMBER_NOT_INT;
+#if !defined(_MSC_VER) || _MSC_VER >= 1900
             if (s == send) {
                 return flags;
             }
-
+#endif
             /* NaN can be followed by various stuff (NaNQ, NaNS), but
              * there are also multiple different NaN values, and some
              * implementations output the "payload" values,
@@ -844,9 +845,11 @@
                 /* C99 style "nan(123)" or Perlish equivalent "nan($uv)". */
                 const char *t;
                 s++;
+#if !defined(_MSC_VER) || _MSC_VER >= 1900
                 if (s == send) {
                     return flags | IS_NUMBER_TRAILING;
                 }
+#endif
                 t = s + 1;
                 while (t < send && *t && *t != ')') {
                     t++;
```
@jameysiddiqui , could you see if that fixes those same issues for you.
Also, try making the same changes to the perl-5.34.0/numeric.c file, too. (I haven't tried that myself, yet.)
Be careful that you place the new preprocessor directives around the correct <code>if(s == send){return...}</code> blocks.
(There's other similar blocks that need to remain as they are.)

This doesn't fix the re/uniprops03.t test failures, and I haven't looked into these in any effective manner, yet. I don't really want to ... but I suppose I might, if I've got nothing better to do.

Is the value I've nominated for _MSC_VER appropriate ? Maybe it needs to be lowered ... I don't know.
What is the value of _MSC_VER with MSVC120 ?

Anyway, if you can verify that this change is helpful to you, we can then find out about whether there's a better solution than the one I've come up with.
First thing to find out is how well it works for you.

**Comment by jameysiddiqui** [comment]
Suggested changes are working fine

Test Summary Report
-------------------
../dist/Net-Ping/t/450_service.t                                   (Wstat: 0 Tests: 26 Failed: 0)
  TODO passed:   9
../ext/IPC-Open3/t/IPC-Open3.t                                     (Wstat: 0 Tests: 45 Failed: 0)
  TODO passed:   25
Files=2677, Tests=1131665, 1637 wallclock secs (97.13 usr +  9.20 sys = 106.33 CPU)
Result: PASS
        cd ..\win32

**Comment by sisyphus** [comment]
Ok - I take it that the changes are working for 5.34.0 ?
If so, I'll make some inquiries about why that troublesome code was added to grok_infnan in 5.32.0.
I'm actually wondering whether that code *ever* does anything useful.

**Comment by hvds** [comment]
> Ok - I take it that the changes are working for 5.34.0 ?
> If so, I'll make some inquiries about why that troublesome code was added to grok_infnan in 5.32.0.
> I'm actually wondering whether that code _ever_ does anything useful.

The first case was added by dca9f615c2 to fix a valgrind error (read past end of string). However it is incorrect, because the docs for grok_infnan promise:
```
If an infinity or a not-a-number is recognized, C<*sp> will point to
one byte past the end of the recognized string.  If the recognition fails,
zero is returned, and C<*sp> will not move.
```
.. so in this case it should be updating `*sp` before returning.

I don't know what's going on with the second block in the patch, tests of `"NaN"` should never be reaching that.

This change should fix the first case:
```
diff --git a/numeric.c b/numeric.c
index a426d97fe5..8fc633046f 100644
--- a/numeric.c
+++ b/numeric.c
@@ -820,6 +820,7 @@ Perl_grok_infnan(pTHX_ const char** sp, const char* send)
 
             flags |= IS_NUMBER_NAN | IS_NUMBER_NOT_INT;
             if (s == send) {
+                *sp = s;
                 return flags;
             }
 
```

**Comment by hvds** [comment]
> This change should fix the first case:

Ah no, looks like it's also supposed to handle trailing whitespace; I'm trying to go through the whole thing to see if there are other problems, and hope to document also precisely what patterns it accepts.

Updated change currently looks like:
```
diff --git a/numeric.c b/numeric.c
index a426d97fe5..7fa9c2802f 100644
--- a/numeric.c
+++ b/numeric.c
@@ -820,7 +820,7 @@ Perl_grok_infnan(pTHX_ const char** sp, const char* send)
 
             flags |= IS_NUMBER_NAN | IS_NUMBER_NOT_INT;
             if (s == send) {
-                return flags;
+                goto ok_trailing;
             }
 
             /* NaN can be followed by various stuff (NaNQ, NaNS), but
@@ -966,6 +966,7 @@ Perl_grok_infnan(pTHX_ const char** sp, const char* send)
             return 0;
     }
 
+  ok_trailing:
     while (s < send && isSPACE(*s))
         s++;
 
```

**Comment by hvds** [comment]
Ugh, I'm an idiot: if we're at the end of the string, clearly there's no trailing whitespace to find.

I did find one more erroneous case, success after "NaNQ" or "NaNS". And much of the rest looks insane, which probably means I don't understand what's going on. The pattern we're matching looks something like:
```
m{
  (
    \+
  |
    -               # IS_NUMBER_NEG
  )?
  (
    (1.#)?          # odh
    (
      inf (         # IS_NUMBER_INFINITY | IS_NUMBER_NOT_INT
        (inity)?
      |
        0+          # only if 'odh' was set above
      )
    |
      ind           # IS_NUMBER_NAN | IS_NUMBER_NOT_INT
    )
    (\s+)           # IS_NUMBER_TRAILING
  |
    [sq] nan (      # IS_NUMBER_NAN | IS_NUMBER_NOT_INT
      [qs]
    |
      \( (          # IS_NUMBER_TRAILING
        # hex, binary or decimal (not octal), that parses as an integer
        # and fits in UV
        $looks_like_number
      ) \)
    |
      (?! \( ) \s* \S  # IS_NUMBER_TRAILING
    )?
  )
  \s+               # IS_NUMBER_TRAILING
}xi
```
I don't understand why we do all the work of parsing the parenthesised part of "NaN(123)" - rejecting some cases, but discarding the result - but then accept any other garbage after "NaN". So eg "NaN(" is a fail, but "NaN[" is a pass.

This is now my best guess at a correct patch:
```
diff --git a/numeric.c b/numeric.c
index a426d97fe5..8032dcb74e 100644
--- a/numeric.c
+++ b/numeric.c
@@ -820,7 +820,7 @@ Perl_grok_infnan(pTHX_ const char** sp, const char* send)
 
             flags |= IS_NUMBER_NAN | IS_NUMBER_NOT_INT;
             if (s == send) {
-                return flags;
+                goto ok_final;
             }
 
             /* NaN can be followed by various stuff (NaNQ, NaNS), but
@@ -834,7 +834,7 @@ Perl_grok_infnan(pTHX_ const char** sp, const char* send)
                  * these portably is tricky. */
                 s++;
                 if (s == send) {
-                    return flags;
+                    goto ok_final;
                 }
             }
             if (*s == '(') {
@@ -972,6 +972,8 @@ Perl_grok_infnan(pTHX_ const char** sp, const char* send)
 #else
     PERL_UNUSED_ARG(send);
 #endif /* #if defined(NV_INF) || defined(NV_NAN) */
+
+  ok_final:
     *sp = s;
     return flags;
 }
```

**Comment by sisyphus** [comment]
S_my_atof_infnan() begins with:
```
    const char *p0 = negative ? s - 1 : s;
    const char *p = p0;
    const int infnan = grok_infnan(&p, send);
    if (infnan && p != p0) {
       /* do stuff ....*/
```
My take is that we want that condition <code>(infnan && p != p0)</code> to be true if and only if the string argument <code>=~ m/^nan|^inf/i</code>
That is currently happening only  for the "inf" strings.
For the "nan" strings, the condition <code>p != p0</code> returns false.

This doesn't produce any errors on most systems, but it does produce an error with these not-so-recent MS compilers because their strtod() implementation is buggy wrt "NaN" strings.

@hvds, your final patch looks good to me but, unfortunately, neither it nor the workaround I provided will work for @jameysiddiqui when (eg) the string is "NaN(" .
He will see:
```
C:\>perl -le "printf '%f', 'nanny';"
NaN
C:\>perl -le "printf '%f', 'nan(';"
0.000000
```
I, too, don't know why the condition <code>((infnan && p != p0)</code> should not hold true when the string is "NaN(".
I'm guessing it has something to do with the nan payload capabilities that @jhi added a few years back.
Maybe we could change that behaviour and see what, if anything, breaks ?
I can probably give that a go some time over the weekend.

**Comment by hvds** [comment]
> @hvds, your final patch looks good to me but, unfortunately, neither it nor the workaround I provided will work for @jameysiddiqui when (eg) the string is "NaN(" .
> He will see:
> 
> ```
> C:\>perl -le "printf '%f', 'nanny';"
> NaN
> C:\>perl -le "printf '%f', 'nan(';"
> 0.000000
> ```

True, but that's not a Windows-specific problem, and not one that causes test failures. On my Ubuntu system:
```
% perl -e 'printf "%f\n", "NaN("'
0.000000
% 
```

So that's a separate issue from what this ticket is about, which is Jamey's test failures.

If you want to make a new issue for "NaN(", feel free; but for me, the bigger surprises are that "NaN[" is happily accepted as "NaN", and the fact that "NaN(" fails but doesn't warn:
```
% perl -we 'printf "%f\n", "NaN("'
0.000000
% perl -we 'printf "%f\n", "NaP("'
Argument "NaP(" isn't numeric in printf at -e line 1.
0.000000
% 
```

Hugo

**Comment by sisyphus** [comment]
@hvds, your Ubuntu perl is apparently significantly different to my Ubuntu perl.
I get:
```
$ perl -wE 'say $]; printf "%f\n", "nan[";'
5.035002
NaN
$ perl -wE 'say $]; printf "%f\n", "nan(";'
5.035002
NaN
```
and it's the same with my mingw-built perls.
And. I believe, if I were to apply your patch, then the outputs I'm seeing on my mingw-built-windows perls and on my ubuntu perls would not change. (Do I need to check this ?)

Does that Ubuntu system you've quoted actually report "NaN" for <code>perl -we 'printf "%f\n", "NaN["'</code> ?
Does the output change when your patch is applied ? 

It seems to me to be pretty obvious that "nan(" should numify to a NaN, as happens universally (?) with the operation <code>"NaN(" + 0</code>:
```
$ perl -wE 'say "NaN(" + 0;'
Argument "NaN(" isn't numeric in addition (+) at -e line 1.
NaN
$ perl -wE 'printf "%f\n", "NaN(" + 0;'
Argument "NaN(" isn't numeric in addition (+) at -e line 1.
NaN
```
And I think it's high time we added 'NaN(' to the list of strings that t/op/infnan.t is testing.

I *will* create a separate issue for this - but not until I can get a handle on what's happening with your Ubuntu system.
Is it failing in the same way as the OP's msvc-built perl, or is it failing via a different pathway ?

**Comment by hvds** [comment]
> @hvds, your Ubuntu perl is apparently significantly different to my Ubuntu perl.

I do apologise, that was with the system perl (5.26); with blead I get the same as you:
```
% ./perl -we 'printf "%f\n", "NaN("'
NaN
% ./perl -we 'printf "%f\n", "NaN["'
NaN
% 
```
both with and without the patch.

I had also been focusing mainly on the "success" part of the docs, and had not concentrated on "If the recognition fails, zero is returned". This appears to have been problematic from the very first implementation (ff4eb3984da), eg when scanning additional letters from "Inf" to "Infinity" we have:
```
                int fail =
                    flags | IS_NUMBER_INFINITY | IS_NUMBER_NOT_INT | IS_NUMBER_TRAILING;
                s++; if (s == send || isALPHA_FOLD_NE(*s, 'N')) return fail;
                s++; if (s == send || isALPHA_FOLD_NE(*s, 'I')) return fail;
...
```
So this is calling it a "fail", and not updating `*sp`, but returning a non-zero result.

As a next step, I'll try looking at the behaviour before ff4eb3984da, when it was still embedded in grok_number.

@jhi if you can recall your intent with the original implementation, I'm sure that would be useful.

**Comment by hvds** [comment]
The original implementation strictly returned zero for every failure case, and fell through to a final `*sp = s; return flags` for every success case. It got a lot of churn after that, though; the first examples of direct non-zero returns started with 1e9aa12fc5 (adding support for "NaN(123)").

It seems to me the original intent was that, eg, "NAND gates" or "infarction" would numify to 0 unless `grok_number_flags` was called with `PERL_SCAN_TRAILING`. That started to change with 8c12dc63d1f, which changed `grok_number_flags` to directly return the result of `grok_infnan`, rather than falling through to the SCAN_TRAILING checks.
 
I feel we should return to honouring `PERL_SCAN_TRAILING`, either by passing that in as a flag to `grok_infnan` or by restoring the checks after calling it in `grok_number_flags`. Choosing the former means changing the interface of a documented API function, which is troublesome; choosing the latter implies that `grok_infnan` should update `*sp` in the case of any non-zero return to point after what was correctly parsed, allowing the caller to check whether it reached `send` or not and react appropriately.

Either way, it leaves open the question of what correct behaviour should be for the other core callers (`isinfnansv` and `my_atof_infnan`), which don't support a SCAN_TRAILING flag.

Getting back to the test failures, my guess is that they are down to this check in `my_atof_infnan`:
```
    const int infnan = grok_infnan(&p, send);
    if (infnan && p != p0) {
```
.. which I think is the only place anyone checks what `grok_infnan` stores in its `*sv`. So I assume that Jamey's platform is a rare case that actually uses this function, and expect that a way of getting those tests to pass, without resolving any of the other issues, is to weaken that test:
```
diff --git a/numeric.c b/numeric.c
index a426d97fe5..cb169ac5bb 100644
--- a/numeric.c
+++ b/numeric.c
@@ -1536,7 +1536,7 @@ S_my_atof_infnan(pTHX_ const char* s, bool negative, const char* send, NV* value
     const char *p0 = negative ? s - 1 : s;
     const char *p = p0;
     const int infnan = grok_infnan(&p, send);
-    if (infnan && p != p0) {
+    if (infnan) {
         /* If we can generate inf/nan directly, let's do so. */
 #ifdef NV_INF
         if ((infnan & IS_NUMBER_INFINITY)) {
```

That's very much papering over the other problems though, so I don't recommend it except at most as a temporary measure.

**Comment by hvds** [comment]
Based on the fact that as far as I can see the intention of `grok_number_flags` has always been to use `PERL_SCAN_TRAILING` to control whether we accept trailing junk or not (defaulting to _not_), my inclination is currently to develop a fix along the following lines:

- in `grok_infnan`, always update `*sp` to point after any _valid_ Inf/NaN string that was parsed; make sure every path through it either returns zero or updates `*sp`
- in `grok_number_flags`, treat an inf/nan parsed with trailing junk as valid only if called with `PERL_SCAN_TRAILING`
- modify `isinfnansv` to disallow trailing junk
- update docs and tests to match this behaviour

And then optionally, as a second patch:
- add support for a new environment variable `PERL_LOOSE_INFNAN`
- modify `my_atof_infnan` to honour that environment variable, so it would stay strict by default but could be made loose (ie permitting trailing junk)
- modify `isinfnansv` to honour that environment variable, as an option to recover the current loose default

I do not propose to modify `grok_number_flags` to look at `PERL_LOOSE_INFNAN`: I think callers to it should be able to rely on it to be strict in the absence of `PERL_SCAN_TRAILING`.

I'd welcome comments, but will also start on the first patch to see what problems it throws up. This will in any case need PSC approval before merging.

**Comment by sisyphus** [comment]
@hvds, there's currently some buggy treatment of the Windows inf/nan symbols ("1.#INF", "1.#IND" and the like).
This, however, is not a windows-specific issue - because other systems also adopt the meaning of these symbols, as ascribed by Windows.
Here is a script that demonstrates (at least some of) those issues:
```
use warnings;
use Test::More;

# If "1.#INF" or "1#INF" is followed by trailing rubbish that =~ m/^i/i
# then we currently get inconsistencies in the way it's handled,
# and also in the issuing of the "isn't numeric" warning.
  
$x = sprintf "%f", "1.#INFi";     # Does not trigger "isn't numeric" warning
$y = sprintf "%f", "1.#INFi" + 0; # Does trigger "isn't numeric" warning
cmp_ok($x + 0, '==', $y + 0, "$x == $y");

$x = sprintf "%f", "-1.#INFI";     # Does not trigger "isn't numeric" warning
$y = sprintf "%f", "-1.#INFI" + 0; # Does trigger "isn't numeric" warning
cmp_ok($x + 0, '==', $y + 0, "$x == $y");

$x = sprintf "%f", "1#INFi";     # Does not trigger "isn't numeric" warning
$y = sprintf "%f", "1#INFi" + 0; # Does trigger "isn't numeric" warning
cmp_ok($x + 0, '==', $y + 0, "$x == $y");

$x = sprintf "%f", "-1#INFI";     # Does not trigger "isn't numeric" warning
$y = sprintf "%f", "-1#INFI" + 0; # Does trigger "isn't numeric" warning
cmp_ok($x + 0, '==', $y + 0, "$x == $y");

# Comments in grok_infnan() seem to imply that "1.#SNAN", "-1.#QNAN",
# and 1.#NAN will numify to a NaN. But that's not always so:

$x = sprintf "%f", "1.#SNAN";
$y = sprintf "%f", "1.#SNAN" + 0;
cmp_ok($x, 'eq', $y, "$x eq $y");

$x = sprintf "%f", "-1.#QNAN";
$y = sprintf "%f", "-1.#QNAN" + 0;
cmp_ok($x, 'eq', $y, "$x eq $y");

$x = sprintf "%f", "1.#NAN";
$y = sprintf "%f", "1.#NAN" + 0;
cmp_ok($x, 'eq', $y, "$x eq $y");

done_testing();
```
On both Windows 7 and Ubuntu-20.04, with perl-5,35,2 && nvtype of double, that script outputs:
```
$ perl infnan.pl
Argument "1.#INFi" isn't numeric in addition (+) at infnan.pl line 9.
not ok 1 - 1.000000 == Inf
#   Failed test '1.000000 == Inf'
#   at infnan.pl line 10.
#          got: 1
#     expected: Inf
Argument "-1.#INFI" isn't numeric in addition (+) at infnan.pl line 13.
not ok 2 - -1.000000 == -Inf
#   Failed test '-1.000000 == -Inf'
#   at infnan.pl line 14.
#          got: -1
#     expected: -Inf
Argument "1#INFi" isn't numeric in addition (+) at infnan.pl line 17.
not ok 3 - 1.000000 == Inf
#   Failed test '1.000000 == Inf'
#   at infnan.pl line 18.
#          got: 1
#     expected: Inf
Argument "-1#INFI" isn't numeric in addition (+) at infnan.pl line 21.
not ok 4 - -1.000000 == -Inf
#   Failed test '-1.000000 == -Inf'
#   at infnan.pl line 22.
#          got: -1
#     expected: -Inf
not ok 5 - 1.000000 eq NaN
#   Failed test '1.000000 eq NaN'
#   at infnan.pl line 29.
#          got: '1.000000'
#     expected: 'NaN'
not ok 6 - -1.000000 eq NaN
#   Failed test '-1.000000 eq NaN'
#   at infnan.pl line 33.
#          got: '-1.000000'
#     expected: 'NaN'
not ok 7 - 1.000000 eq NaN
#   Failed test '1.000000 eq NaN'
#   at infnan.pl line 37.
#          got: '1.000000'
#     expected: 'NaN'
1..7
# Looks like you failed 7 tests of 7.
```
Amazingly, when we switch to nvtypes of "long double" or "__float128" (same version of perl and same systems), all tests pass, with Infs and NaNs being assigned as expected, and the only issue is the absence of warnings as described in the comments of the script.

Apart perhaps for the failing tests 5, 6 and 7, I don't think these bugs are of great significance. For me, the most disconcerting thing about them is the underlying confusion of which they're indicative.

**Comment by hvds** [comment]
I've pushed the branch hv/gh19010  with a single commit (a9475d1841) implementing part of my proposal above:
* in `grok_infnan`, always update `*sp` to point after any valid Inf/NaN string that was parsed; make sure every path through it either returns zero or updates *sp`
* in grok_number_flags, treat an inf/nan parsed with trailing junk as valid only if called with PERL_SCAN_TRAILING
* update docs and tests to match this behaviour

I've not modified `isinfnansv` to be strict, but added a caveat to its docs.

Happily (or not), tests in `op/infnan.t` all pass without modification; a couple of tests in `XS-APItest/t/grok.t` needed changes.

@sisyphus I think there's a good chance this branch will also clean up the inconsistencies shown by your test script.

@jhi, @rjbs, @neilb, @leonerd Since this fix requires adaptive test changes, I'd like to know whether PSC is in principle ok with the direction of travel before I make a PR for it. I'd also love to get input from Jarkko.

**Comment by sisyphus** [comment]
LGTM.

In <code>S_my_atof_infnan()</code> the condition <code>(infnan && p != p0)</code> is now returning as per expected.
I think this will therefore fix the original issue that @jameysiddiqui has reported here, though I'm not set up to test this conclusively.

**Comment by hvds** [comment]
> I've not modified `isinfnansv` to be strict, but added a caveat to its docs.

I note that if I do so modify isinfnansv, no tests fail.

The only tests that appear to exercise it at all are `op/infnan.t` 43, 50-51, 58-59, 62-63, 359-360, 363-364, 367-368 (checked by modifying it to return false always). So if we want to assert its behaviour one way or the other, duplicating some of those tests would be the obvious way to get there.

**Comment by hvds** [comment]
I've created #19138 for this, but unsure who is likely to review it.

**Comment by hvds** [comment]
#19138 has now been merged, so this should be fixed in v5.36.
